### PR TITLE
Razer Tartarus v2 Lighting Support

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by z3ntu** [comment]
What's the state of this PR?

**Comment by ghost** [comment]
@z3ntu I resolved the merge conflicts.

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 1]
```diff
@@ -176,6 +176,54 @@ def _close(self):
...
+               'get_macro_effect', 'set_macro_effect', 'get_macros', 'delete_macro', 'add_macro',
+               'set_ripple_effect', 'set_ripple_effect_random_colour',
+               'get_game_mode', 'set_game_mode',
+               'keypad_get_mode_modifier', 'keypad_set_mode_modifier']
+    DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/1255/1255_tartarus_v2.png"
+
+    RAZER_URLS = {
+        "top_img": "https://d4kkpd69xt9l7.cloudfront.net/sys-master/root/h96/he9/8909630603294/razer-tartarus-v2-gallery-08.jpg",
+        "side_img": "https://d4kkpd69xt9l7.cloudfront.net/sys-master/root/h44/hea/8909630668830/razer-tartarus-v2-gallery-09.jpg",
+        "perspective_img": "https://d4kkpd69xt9l7.cloudfront.net/sys-master/root/h56/hed/8909630734366/razer-tartarus-v2-gallery-10.jpg"
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: If you could find any picture from the support.razer.com Domain, that would be better :) If not we can leave those

File: driver/razerkbd_driver.c [hunk 2]
```diff
@@ -865,6 +872,10 @@ static ssize_t razer_attr_write_mode_spectrum(struct device *dev, struct device_
         razer_send_payload(usb_dev, &report);
         report = razer_chroma_standard_set_led_effect(VARSTORE, BACKLIGHT_LED, LED_SPECTRUM_CYCLING);
         break;
+    case USB_DEVICE_ID_RAZER_TARTARUS_V2:
+        report = razer_chroma_extended_matrix_effect_spectrum(VARSTORE, BACKLIGHT_LED);
+        report.transaction_id.id = 0x1F;  // TODO move to a usb_device variable
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:886]: newline before and after the new text please

File: driver/razerkbd_driver.c [hunk 3]
```diff
@@ -935,6 +947,11 @@ static ssize_t razer_attr_write_mode_static(struct device *dev, struct device_at
     struct razer_report report = {0};
 
     switch(usb_dev->descriptor.idProduct) {
+    case USB_DEVICE_ID_RAZER_TARTARUS_V2:
+        report = razer_chroma_extended_matrix_effect_static(VARSTORE, BACKLIGHT_LED, (struct razer_rgb*)&buf[0]);
+        razer_send_payload(usb_dev, &report);
+        report.transaction_id.id = 0x1F;
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:965]: same

File: driver/razerkbd_driver.c [hunk 4]
```diff
@@ -1410,6 +1472,10 @@ static ssize_t razer_attr_write_set_brightness(struct device *dev, struct device
     struct razer_report report = {0};
 
     switch(usb_dev->descriptor.idProduct) {
+    case USB_DEVICE_ID_RAZER_TARTARUS_V2:
+        report = razer_chroma_extended_matrix_brightness(VARSTORE, ZERO_LED, brightness);
+        report.transaction_id.id = 0x1F;
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:1493]: same

File: driver/razerkbd_driver.c [hunk 5]
```diff
@@ -1460,6 +1526,10 @@ static ssize_t razer_attr_read_set_brightness(struct device *dev, struct device_
     struct razer_report response = {0};
 
     switch(usb_dev->descriptor.idProduct) {
+    case USB_DEVICE_ID_RAZER_TARTARUS_V2:
+        report = razer_chroma_extended_matrix_get_brightness(VARSTORE, ZERO_LED);
+        report.transaction_id.id = 0x1F;
+        break;
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:1550]: same

File: driver/razerkbd_driver.c [hunk 6]
```diff
@@ -2121,6 +2192,34 @@ static int razer_kbd_probe(struct hid_device *hdev, const struct hid_device_id *
...
+        case USB_DEVICE_ID_RAZER_ORBWEAVER:
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_effect_static);          // Static effect
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_effect_pulsate);         // Pulsate effect, like breathing
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_matrix_effect_none);            // No effect
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_profile_led_red);               // Profile/Macro LED Red
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_profile_led_green);             // Profile/Macro LED Green
+            CREATE_DEVICE_FILE(&hdev->dev, &dev_attr_profile_led_blue);              // Profile/Macro LED Blue
+            break;
+
+        case USB_DEVICE_ID_RAZER_ORBWEAVER_CHROMA:
```
Comment (z3ntu) [inline driver/razerkbd_driver.c]: I think a merge conflict was resolved wrong here (and below)

**Review by z3ntu — CHANGES_REQUESTED** [review]
See comments, some minor nitpicks

**Comment by z3ntu** [comment]
Also do all the lighting effects work properly? Is it working well?

**Comment by ghost** [comment]
I no longer have this device. Every working lighting function was mentioned in the related issue.

**Comment by ghost** [comment]
Would you please use a new format to generate the `install_files/udev/99-razer.rules` file from? Every time someone will create a new patch this file will conflict.

File: daemon/openrazer_daemon/hardware/keyboards.py (hunk 1 above)
Comment (ghost) [reply daemon/openrazer_daemon/hardware/keyboards.py]: Couldn't find anything.

**Review by z3ntu — APPROVED** [review]
LGTM, thank you for the PR!

**Comment by lukastheblack** [comment]
Was this ever merged? I am trying to configure my tartarus v2 on Arch, and it is recognized as a keyboard.... but not recognized by the openrazer-daemon. Let me know if there is anything I can do to help. thank you.

EDIT: I noticed today that in the repository, the device is identified as 1532:022B, but my device is 1532:0244

Trying to figure out the correct way to build this locally on Arch, but the main page only speaks of installing it from AUR. Let me know if you can provide any guidance and I will test my changes locally.

**Comment by nightsky30** [comment]
> EDIT: I noticed today that in the repository, the device is identified as 1532:022B, but my device is 1532:0244

@lukastheblack 1532:0244 is the PID for the Tartarus Pro, which is not supported yet. You can check the status with issue #1039.

**Pull request merged**